### DISCUSSION TITLE
Shared WDIO config and various WDIO browser config files

### DISCUSSION THREAD

**Discussion by Songithub95** [webdriverio/webdriverio#14003] [Q&A]
Hi folks ! I need sample of complete cuztomized config files with full of parameters which are shared.conf.js file (this file is connected to other separate browser config files), chrome.conf.js file, safari.conf.js file , firefox.conf.js file. I've already looked into the documentation but it is just a reference of a small part of a particular browser config. Anyone has any advice ? Many thanks up front !

**Comment by portyu9** [comment]
Hi @Songithub95 

You generally **shouldn't create a config containing every WebdriverIO option**. Most options already have sensible defaults, so a shared base config containing only the settings your project actually changes is easier to maintain.

A clean pattern is:

```ts
// wdio.shared.conf.ts
import type { Options } from '@wdio/types'

export const sharedConfig: Options.Testrunner = {
    runner: 'local',

    specs: [
        './test/specs/**/*.ts'
    ],

    maxInstances: 5,
    logLevel: 'info',

    framework: 'mocha',

    reporters: ['spec'],

    waitforTimeout: 10000,
    connectionRetryTimeout: 120000,
    connectionRetryCount: 3,

    mochaOpts: {
        ui: 'bdd',
        timeout: 60000
    }
}
```

Then keep only browser-specific capabilities in each browser config.

```ts
// wdio.chrome.conf.ts
import type { Options } from '@wdio/types'
import { sharedConfig } from './wdio.shared.conf.js'

export const config: Options.Testrunner = {
    ...sharedConfig,

    capabilities: [{
        browserName: 'chrome'
    }]
}
```

```ts
// wdio.firefox.conf.ts
import type { Options } from '@wdio/types'
import { sharedConfig } from './wdio.shared.conf.js'

export const config: Options.Testrunner = {
    ...sharedConfig,

    capabilities: [{
        browserName: 'firefox'
    }]
}
```

```ts
// wdio.safari.conf.ts
import type { Options } from '@wdio/types'
import { sharedConfig } from './wdio.shared.conf.js'

export const config: Options.Testrunner = {
    ...sharedConfig,

    capabilities: [{
        browserName: 'safari'
    }]
}
```

Then run the appropriate config, for example:

```bash
npx wdio run ./wdio.chrome.conf.ts
```

One thing to remember is that JavaScript object spreading is **shallow**. If both configs need to modify a nested object such as `mochaOpts`, merge that property explicitly:

```ts
mochaOpts: {
    ...sharedConfig.mochaOpts,
    timeout: 90000
}
```

The WebdriverIO configuration reference is intended to document the complete option set; your own config should normally contain only the options you need to override.

Please mark this answer as accepted if it helped, thank you.
